README.md
[![Build Status](https://travis-ci.org/akzhan/perl-Plack-Session-Store-RedisFast.svg?branch=master)](https://travis-ci.org/akzhan/perl-Plack-Session-Store-RedisFast)
[![codecov](https://codecov.io/gh/akzhan/perl-Plack-Session-Store-RedisFast/branch/master/graph/badge.svg)](https://codecov.io/gh/akzhan/perl-Plack-Session-Store-RedisFast)

# NAME

Plack::Session::Store::RedisFast - Redis session store.

# DESCRIPTION

Default implementation of Redis handle is [Redis::Fast](https://metacpan.org/pod/Redis::Fast); otherwise [Redis](https://metacpan.org/pod/Redis).

May be overriden through ["redis"](#redis) param.

Default implementation of serializer handle is [JSON::XS](https://metacpan.org/pod/JSON::XS); otherwise [Mojo::JSON](https://metacpan.org/pod/Mojo::JSON) or [JSON](https://metacpan.org/pod/JSON).

May be overriden through ["inflate"](#inflate) and ["deflate"](#deflate) param.

# SYNOPSIS

    use Plack::Builder;
    use Plack::Session::Store::RedisFast;

    my $app = sub {
        return [ 200, [ 'Content-Type' => 'text/plain' ], [ 'Hello Foo' ] ];
    };

    builder {
        enable 'Session',
            store => Plack::Session::Store::RedisFast->new;
        $app;
    };

# DESCRIPTION

This will persist session data using [Redis::Fast](https://metacpan.org/pod/Redis::Fast) or [Redis](https://metacpan.org/pod/Redis).

This is a subclass of [Plack::Session::Store](https://metacpan.org/pod/Plack::Session::Store) and implements
its full interface.

# METHODS

## new

    Plack::Session::Store::RedisFast->new( %param );

Parameters:

- redis

    A simple accessor for the Redis handle.

- inflate

    A simple serializer, requires ["deflate"](#deflate) param.

- deflate

    A simple deserializer, requires ["inflate"](#inflate) param.

- encoder

    A simple encoder (encode/decode implementation), class or instance. JSON/utf8 by default.

- prefix

    A prefix for Redis session ids. 'Plack::Session::Store::RedisFast:' by default.

- expires

    An expires for Redis sessions. ["ONE\_MONTH" in Time::Seconds](https://metacpan.org/pod/Time::Seconds#ONE_MONTH) by default.

## each\_session

    $store->each_session(sub {
        my ( $redis_instance, $redis_prefix, $session_id, $session ) = @_;
    });

Enumerates all stored sessions using SCAN, see [https://redis.io/commands/scan](https://redis.io/commands/scan) for limitations.

# BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please fill in GitHub issues.

# AUTHOR

Akzhan Abdulin <akzhan.abdulin@gmail.com>

# COPYRIGHT

Copyright 2017- Akzhan Abdulin

# LICENSE

MIT License


Changes
Revision history for perl-Plack-Session-Store-RedisFast

{{$NEXT}}

0.05  2017-06-26 07:55:54 MSK

        - expire param renamed to expires.

0.04  2017-06-25 05:35:50 MSK

        - builder param replaced with redis one.

0.03  2017-06-11 17:21:33 MSK

        - minor update.

0.02  2017-06-11 17:12:09 MSK

        - each_session method introduced.

        - support for Mojo::JSON encoder added.

0.01  2017-06-07 11:25:56 MSK

        - Initial release


lib/Plack/Session/Store/Redis.pm
package Plack::Session::Store::Redis;

use strict;
use warnings;

use 5.008_005;

use Carp qw( carp );
use Plack::Util::Accessor qw( prefix redis inflate deflate expire );
use Time::Seconds qw( ONE_MONTH );

use parent 'Plack::Session::Store';

our $VERSION   = '0.01';
our $AUTHORITY = 'cpan:AKZHAN';

sub new {
    my ( $class, %param ) = @_;
    $param{prefix} = __PACKAGE__ . ':' unless defined $param{prefix};
    $param{expire} = ONE_MONTH         unless exists $param{expire};

    $param{inflate} ||= \&_inflate;
    $param{deflate} ||= \&_deflate;

    unless ( $param{redis} ) {
        my $builder = $param{builder} || \&_build_redis;
        delete $param{builder};
        $param{redis} = $builder->();
    }

    bless {%param} => $class;
}

sub _build_redis {
    my $instance;
    eval {
        require Redis::Fast;
        $instance = Redis::Fast->new;
        1;
    } or do {
        require Redis;
        $instance = Redis->new;
    };
    $instance;
}

sub _build_encoder {
    my $instance;
    eval {
        require JSON::XS;
        $instance = JSON::XS->new->utf8->allow_nonref;
        1;
    } or do {
        require JSON;
        $instance = JSON->new->utf8->allow_nonref;
    };
    $instance;
}

my $_encoder = undef;

sub _encoder {
    $_encoder ||= _build_encoder();
}

sub _inflate {
    my ($session) = @_;
    _encoder->encode($session);
}

sub _deflate {
    my ($data) = @_;
    _encoder->decode($data);
}

sub fetch {
    my ( $self, $session_id ) = @_;
    my $data = $self->redis->get( $self->prefix . $session_id );
    return undef unless defined $data;
    $self->deflate->($data);
}

sub store {
    my ( $self, $session_id, $session ) = @_;
    unless ( defined $session ) {
        carp "store: no session provided";
        return;
    }
    my $data = $self->inflate->($session);
    $self->redis->set(
        $self->prefix . $session_id => $data,
        ( defined( $self->expire ) ? ( EX => $self->expire ) : () ),
    );
}

sub remove {
    my ( $self, $session_id ) = @_;
    $self->redis->del( $self->prefix . $session_id );
}

1;

__END__

=pod

=encoding utf-8


=head1 NAME

Plack::Session::Store::Redis - Redis session store

=head1 SYNOPSIS

  use Plack::Builder;
  use Plack::Session::Store::Redis;

  my $app = sub {
      return [ 200, [ 'Content-Type' => 'text/plain' ], [ 'Hello Foo' ] ];
  };

  builder {
      enable 'Session',
          store => Plack::Session::Store::Redis->new;
      $app;
  };

=head1 DESCRIPTION

This will persist session data using L<Redis::Fast> or L<Redis>.

This is a subclass of L<Plack::Session::Store> and implements
its full interface.

=head1 METHODS

=over 4

=item B<new ( %params )>

=item B<redis>

A simple accessor for the Redis handle.

=back

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please fill in GitHub issues.

=head1 AUTHOR

Akzhan Abdulin E<lt>akzhan.abdulin@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2017- Akzhan Abdulin

=head1 LICENSE

MIT License

=cut


lib/Plack/Session/Store/RedisFast/MojoJSON.pm
package Plack::Session::Store::RedisFast::MojoJSON;

use strict;
use warnings;

use 5.008_005;

use Mojo::JSON qw( decode_json encode_json );

sub new {
    my ($class) = @_;

    bless {}, $class;
}

sub encode {
    my ( $self, $thing ) = @_;

    encode_json($thing);
}

sub decode {
    my ( $self, $bytes ) = @_;

    decode_json($bytes);
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Plack::Session::Store::RedisFast::MojoJSON - Mojo::JSON adapte for Redis session store.

=head1 DESCRIPTION

L<Mojo::JSON>-based implementation of L<Plack::Session::Store::RedisFast/inflate>
and L<Plack::Session::Store::RedisFast/deflate>.

=head1 SYNOPSIS

    use Plack::Session::Store::RedisFast::MojoJSON;

    my $encoder = Plack::Session::Store::RedisFast::MojoJSON->new;

    my $bytes = $encoder->encode( $hashref );

    my $hashref2 = $encoder->decode( $bytes );

=head1 DESCRIPTION

Used by default when L<JSON::XS> is not available.

=head1 METHODS

=head2 new

    Plack::Session::Store::RedisFast::MojoJSON->new;

=cut


lib/Plack/Session/Store/RedisFast/Mojo/JSON.pm
package Plack::Session::Store::RedisFast::Mojo::JSON;

use strict;
use warnings;

use 5.008_005;

use Mojo::JSON qw( decode_json encode_json );

sub new {
    my ($class) = @_;

    bless {}, $class;
}

sub encode {
    my ( $self, $thing ) = @_;

    encode_json($thing);
}

sub decode {
    my ( $self, $bytes ) = @_;

    decode_json($bytes);
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Plack::Session::Store::RedisFast::MojoJSON - Mojo::JSON adapte for Redis session store.

=head1 DESCRIPTION

L<Mojo::JSON>-based implementation of L<Plack::Session::Store::RedisFast/inflate>
and L<Plack::Session::Store::RedisFast/deflate>.

=head1 SYNOPSIS

    use Plack::Session::Store::RedisFast::MojoJSON;

    my $encoder = Plack::Session::Store::RedisFast::MojoJSON->new;

    my $bytes = $encoder->encode( $hashref );

    my $hashref2 = $encoder->decode( $bytes );

=head1 DESCRIPTION

Used by default when L<JSON::XS> is not available.

=head1 METHODS

=head2 new

    Plack::Session::Store::RedisFast::MojoJSON->new;

=cut


lib/Plack/Session/Store/RedisFast.pm
package Plack::Session::Store::RedisFast;

use strict;
use warnings;

use 5.008_005;

use Carp qw( carp );
use Plack::Util::Accessor qw( prefix redis encoder expires );
use Time::Seconds qw( ONE_MONTH );

use parent 'Plack::Session::Store';

use constant SESSIONS_PER_SCAN => 100;

our $VERSION   = '0.04';
our $AUTHORITY = 'cpan:AKZHAN';

sub new {
    my ( $class, %param ) = @_;
    $param{prefix} = __PACKAGE__ . ':' unless defined $param{prefix};
    if ( exists $param{expire} ) {
        warn __PACKAGE__
          . ": DEPRECATED expire. Should be replaced with expires.";
        $param{expires} = delete $param{expire};
    }
    $param{expires} = ONE_MONTH unless exists $param{expires};

    unless ( defined $param{redis} ) {
        $param{redis} = \&_build_redis;
    }

    if ( ref( $param{redis} ) eq 'CODE' ) {
        $param{redis} = $param{redis}->();
    }

    $param{encoder} ||=
      _build_encoder( ( delete $param{inflate} ), ( delete $param{deflate} ) );

    $param{encoder} = $param{encoder}->new()
      unless ref( $param{encoder} );

    bless {%param} => $class;
}

sub _build_redis {
    my $instance;
    eval {
        require Redis::Fast;
        $instance = Redis::Fast->new;
        1;
    } or do {
        require Redis;
        $instance = Redis->new;
    };
    $instance;
}

sub _build_encoder {
    my ( $inflate, $deflate ) = @_;
    if ( $inflate && $deflate ) {
        require Plack::Session::Store::RedisFast::Encoder::Custom;
        return Plack::Session::Store::RedisFast::Encoder::Custom->new( $inflate,
            $deflate );
    }
    my $instance;
    eval {
        require Plack::Session::Store::RedisFast::Encoder::JSONXS;
        $instance = Plack::Session::Store::RedisFast::Encoder::JSONXS->new;
        1;
    } or do {
        require Plack::Session::Store::RedisFast::Encoder::MojoJSON;
        $instance = Plack::Session::Store::RedisFast::Encoder::MojoJSON->new;
      }
      or do {
        require Plack::Session::Store::RedisFast::Encoder::JSON;
        $instance = Plack::Session::Store::RedisFast::Encoder::JSON->new;
      };
    $instance;
}

sub fetch {
    my ( $self, $session_id ) = @_;
    my $data = $self->redis->get( $self->prefix . $session_id );
    return undef unless defined $data;
    $self->encoder->decode($data);
}

sub store {
    my ( $self, $session_id, $session ) = @_;
    unless ( defined $session ) {
        carp "store: no session provided";
        return;
    }
    my $data = $self->encoder->encode($session);
    $self->redis->set(
        $self->prefix . $session_id => $data,
        ( defined( $self->expires ) ? ( EX => $self->expires ) : () ),
    );
    1;
}

sub remove {
    my ( $self, $session_id ) = @_;
    $self->redis->del( $self->prefix . $session_id );
    1;
}

sub each_session {
    my ( $self, $cb ) = @_;
    return if ref($cb) ne 'CODE';

    my $prefix = $self->prefix;

    my $cursor = 0;
    for ( ; ; ) {
        ( $cursor, my $keys ) = $self->redis->scan(
            $cursor,
            MATCH => $self->prefix . '*',
            COUNT => SESSIONS_PER_SCAN
        );
        if ( scalar(@$keys) > 0 ) {
            my @sessions = $self->redis->mget(@$keys);

            for ( my $i = 0 ; $i < scalar(@sessions) ; $i++ ) {
                next unless $sessions[$i];

                next if $keys->[$i] !~ m/^\Q$prefix\E(.+)$/;
                my $session_id = $1;

                $cb->(
                    $self->redis, $prefix, $session_id,
                    $self->encoder->decode( $sessions[$i] ),
                );
            }
        }

        last if $cursor == 0;
    }
    1;
}

1;

__END__

=pod

=encoding utf-8

=head1 NAME

Plack::Session::Store::RedisFast - Redis session store.

=head1 DESCRIPTION

Default implementation of Redis handle is L<Redis::Fast>; otherwise L<Redis>.

May be overriden through L</redis> param.

Default implementation of serializer handle is L<JSON::XS>; otherwise L<Mojo::JSON> or L<JSON>.

May be overriden through L</inflate> and L</deflate> param.

=head1 SYNOPSIS

  use Plack::Builder;
  use Plack::Session::Store::RedisFast;

  my $app = sub {
      return [ 200, [ 'Content-Type' => 'text/plain' ], [ 'Hello Foo' ] ];
  };

  builder {
      enable 'Session',
          store => Plack::Session::Store::RedisFast->new;
      $app;
  };

=head1 DESCRIPTION

This will persist session data using L<Redis::Fast> or L<Redis>.

This is a subclass of L<Plack::Session::Store> and implements
its full interface.

=head1 METHODS

=head2 new

    Plack::Session::Store::RedisFast->new( %param );

Parameters:

=over 4

=item redis

A simple accessor for the Redis handle.

=item inflate

A simple serializer, requires L</deflate> param.

=item deflate

A simple deserializer, requires L</inflate> param.

=item encoder

A simple encoder (encode/decode implementation), class or instance. JSON/utf8 by default.

=item prefix

A prefix for Redis session ids. 'Plack::Session::Store::RedisFast:' by default.

=item expires

An expires for Redis sessions. L<Time::Seconds/ONE_MONTH> by default.

=back

=head2 each_session

    $store->each_session(sub {
        my ( $redis_instance, $redis_prefix, $session_id, $session ) = @_;
    });

Enumerates all stored sessions using SCAN, see L<https://redis.io/commands/scan> for limitations.

=head1 BUGS

All complex software has bugs lurking in it, and this module is no
exception. If you find a bug please fill in GitHub issues.

=head1 AUTHOR

Akzhan Abdulin E<lt>akzhan.abdulin@gmail.comE<gt>

=head1 COPYRIGHT

Copyright 2017- Akzhan Abdulin

=head1 LICENSE

MIT License

=cut
